[dtensor][debug] have visualize_sharding correctly print for sub-mesh DTensor

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120265
* #121217
* __->__ #121216
* #120260
* #121179

**Summary**
In `visualize_sharding` we chose to only print on rank 0 (global rank) which means calling `visualize_sharind` will never print anything when the dtensor object's mesh doesn't include rank 0 (i.e. a sub-mesh). This PR has `visualize_sharding` always print on rank whose mesh coordinate is (0, 0, ..., 0) instead of whose global rank is 0.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

XilunWu committed:
[dtensor][debug] have visualize_sharding correctly print for sub-mesh DTensor

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/debug/visualize_sharding.py:
```diff
@@ -153,6 +153,10 @@ def visualize_sharding(dtensor, prompt=""):
     blocks = _convert_offset_to_ranges(all_offsets)
-    if device_mesh.get_rank() == 0:
+    local_rank_zero_on_all_dim = all(
+        device_mesh.get_local_rank(mesh_dim=dim) == 0
```
I think we can simply do this?
`device_mesh.get_local_rank(mesh_dim=0) == 0`

XilunWu replied:
this may not be correct for 2-d mesh. For example, saying the mesh is [[0, 1], [2, 3]]. `get_local_rank(mesh_dim=0)  == 0` is `True` on rank 0 and 1 whose first coordinate is 0.

wanchaol replied:
Ohh i see

wanchaol reviewed (approved):
lgtm!

XilunWu committed:
Update on "[dtensor][debug] have visualize_sharding correctly print for sub-mesh DTensor"


**Summary**
In `visualize_sharding` we chose to only print on rank 0 (global rank) which means calling `visualize_sharind` will never print anything when the dtensor object's mesh doesn't include rank 0 (i.e. a sub-mesh). This PR has `visualize_sharding` always print on rank whose mesh coordinate is (0, 0, ..., 0) instead of whose global rank is 0.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][debug] have visualize_sharding correctly print for sub-mesh DTensor"


**Summary**
In `visualize_sharding` we chose to only print on rank 0 (global rank) which means calling `visualize_sharind` will never print anything when the dtensor object's mesh doesn't include rank 0 (i.e. a sub-mesh). This PR has `visualize_sharding` always print on rank whose mesh coordinate is (0, 0, ..., 0) instead of whose global rank is 0.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@pytorchbot merge -f "unrelated test failure"
